File: src-tauri/src/application/flow_service.rs
use std::sync::Arc;

use crate::{
    application::ports::{FlowRepository, PortError},
    domain::Flow,
};

pub struct FlowService<R: FlowRepository + ?Sized> {
    repository: Arc<R>,
}

impl<R: FlowRepository + ?Sized> FlowService<R> {
    pub fn new(repository: Arc<R>) -> Self {
        Self { repository }
    }

    pub async fn save_flow(&self, flow: &Flow) -> Result<Flow, PortError> {
        self.repository.save_flow(flow).await?;
        self.repository
            .load_flow(&flow.id)
            .await?
            .ok_or_else(|| PortError::new("FLOW_NOT_FOUND", "flow could not be reloaded"))
    }

    pub async fn list_flows(&self) -> Result<Vec<Flow>, PortError> {
        self.repository.list_flows().await
    }

    pub async fn duplicate_flow(
        &self,
        flow_id: &str,
        duplicate_id: &str,
    ) -> Result<Flow, PortError> {
        if self.repository.load_flow(duplicate_id).await?.is_some() {
            return Err(PortError::new(
                "FLOW_ALREADY_EXISTS",
                "flow ID already exists",
            ));
        }
        let mut duplicate = self
            .repository
            .load_flow(flow_id)
            .await?
            .ok_or_else(|| PortError::new("FLOW_NOT_FOUND", "flow not found"))?;
        duplicate.id = duplicate_id.into();
        duplicate.name = format!("{} copy", duplicate.name);
        self.save_flow(&duplicate).await
    }
}

File: electron/application/flowService.ts
import type { ConnectionProfile, Flow } from "../domain/models";
import {
  SqlValidationError,
  validateSourceStatement,
  validateTargetStatement,
} from "../domain/sqlValidation";
import type { FlowRepository } from "./ports";

export type { FlowRepository } from "./ports";

export class FlowServiceError extends Error {
  constructor(public readonly code: string, message: string) {
    super(message);
    this.name = "FlowServiceError";
  }
}

export class FlowService {
  constructor(private readonly repository: FlowRepository) {}

  async saveFlow(flow: Flow): Promise<Flow> {
    const normalized = normalizeFlowStepTitles(flow);
    validateFlow(normalized);
    if (this.repository.loadConnection(normalized.sourceConnectionId) === undefined) {
      throw new FlowServiceError(
        "CONNECTION_NOT_FOUND",
        "source connection was not found",
      );
    }
    if (this.repository.loadConnection(normalized.targetConnectionId) === undefined) {
      throw new FlowServiceError(
        "CONNECTION_NOT_FOUND",
        "target connection was not found",
      );
    }
    this.repository.saveFlow(normalized);
    const saved = this.repository.loadFlow(normalized.id);
    if (saved === undefined) {
      throw new FlowServiceError("FLOW_NOT_FOUND", "flow could not be reloaded");
    }
    return saved;
  }

  async listFlows(): Promise<Flow[]> {
    return this.repository.listFlows();
  }

  async deleteFlow(flowId: string): Promise<void> {
    validateRequired(flowId, "flow ID");
    if (this.repository.loadFlow(flowId) === undefined) {
      throw new FlowServiceError("FLOW_NOT_FOUND", "flow not found");
    }
    this.repository.deleteFlow(flowId);
  }

  async duplicateFlow(flowId: string, duplicateId: string): Promise<Flow> {
    validateRequired(flowId, "flow ID");
    validateRequired(duplicateId, "duplicate flow ID");
    if (this.repository.loadFlow(duplicateId) !== undefined) {
      throw new FlowServiceError("FLOW_ALREADY_EXISTS", "flow ID already exists");
    }
    const source = this.repository.loadFlow(flowId);
    if (source === undefined) {
      throw new FlowServiceError("FLOW_NOT_FOUND", "flow not found");
    }
    return this.saveFlow({
      ...structuredClone(source),
      id: duplicateId,
      name: `${source.name} copy`,
    });
  }
}

export function normalizeFlowStepTitles(flow: Flow): Flow {
  return {
    ...flow,
    querySteps: flow.querySteps.map((step, position) => ({
      ...step,
      title: step.title?.trim() || `Step ${position + 1}`,
    })),
  };
}

function validateFlow(flow: Flow): void {
  validateRequired(flow.id, "flow ID");
  validateRequired(flow.name, "flow name");
  validateRequired(flow.sourceConnectionId, "source connection ID");
  validateRequired(flow.targetConnectionId, "target connection ID");
  if (flow.querySteps.length === 0) {
    throw new FlowServiceError("VALIDATION", "at least one query step is required");
  }
  for (const step of flow.querySteps) {
    validateRequired(step.id, "query step ID");
    validateRequired(step.selectSql, "source SQL");
    validateRequired(step.upsertSql, "target SQL");
    try {
      validateSourceStatement(step.selectSql);
      validateTargetStatement("oracle", step.upsertSql);
    } catch (error) {
      if (error instanceof SqlValidationError) {
        throw new FlowServiceError("VALIDATION", error.message);
      }
      throw error;
    }
  }
}

function validateRequired(value: string, label: string): void {
  if (value.trim().length === 0) {
    throw new FlowServiceError("VALIDATION", `${label} is required`);
  }
}
